src/compiler/checker.ts

```diff
@@ -12073,7 +12073,20 @@ namespace ts {
         }
 
         function isGenericMappedType(type: Type): type is MappedType {
-            return !!(getObjectFlags(type) & ObjectFlags.Mapped) && isGenericIndexType(getConstraintTypeFromMappedType(type as MappedType));
+            if (getObjectFlags(type) & ObjectFlags.Mapped) {
+                const constraint = getConstraintTypeFromMappedType(type as MappedType);
+                if (isGenericIndexType(constraint)) {
+                    return true;
+                }
+                // A mapped type is generic if the as clause references other generic types than the iteration type.
+                // To determine this, we substitute the constraint type (that we now know isn't generic) for the iteration

```

Comment:
If you're ever mapping on an empty type, you will end up with `never` in some cases, which will make the mapped type not generic. This means that if an interface is intended to be merged with other interfaces, this will return different answers pre-merged an post-merged. At the very least, we should have a test for that.

Otherwise, should we consider just instantiating the mapped type variable with the string literal type `""` instead?